src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -330,18 +342,43 @@ private static int VerifyClientCertificate(int preverify_ok, IntPtr x509_ctx_ptr
             return OpenSslSuccess;
         }
 
-        private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out IntPtr outp, out byte outlen, IntPtr inp, uint inlen, IntPtr arg)
+        private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out byte* outp, out byte outlen, byte* inp, uint inlen, IntPtr arg)
         {
-            outp = IntPtr.Zero;
+            outp = null;
             outlen = 0;
-                        
-            GCHandle protocols = GCHandle.FromIntPtr(arg);
-            Debug.Assert(protocols.IsAllocated && protocols.Target != null);
 
-            byte[] server = (byte[])protocols.Target;
+            GCHandle protocolHandle = GCHandle.FromIntPtr(arg);
+            if (!(protocolHandle.Target is List<SslApplicationProtocol> protocolList))
+            {
+                return Ssl.SSL_TLSEXT_ERR_NOACK;
+            }
+
+            try
+            {
+                for (int i = 0; i < protocolList.Count; i++)
+                {
+                    Span<byte> clientList = new Span<byte>(inp, (int)inlen);
+                    while (clientList.Length > 0)
+                    {
+                        byte length = clientList[0];
+                        Span<byte> clientProto = clientList.Slice(1, length);
+                        if (clientProto.SequenceEqual(protocolList[i].Protocol.Span))
+                        {
+                            outp = inp + (int)inlen - clientList.Length + 1;
+                            outlen = length;
+                            return Ssl.SSL_TLSEXT_ERR_OK;
+                        }
+
+                        clientList = clientList.Slice(1 + length);
+                    }
+                }
+            }
+            catch
+            {
+                return Ssl.SSL_TLSEXT_ERR_NOACK;
+            }
 
-            return Ssl.SslSelectNextProto(out outp, out outlen, protocols.AddrOfPinnedObject(), (uint)server.Length, inp, inlen) == Ssl.OPENSSL_NPN_NEGOTIATED ?
```

Comment:
Why was this previously called, and now isn't?

Reply:
There is a problem with that is it returns a pointer inside either the client or server buffer. The client buffer is fine because it's created and managed by openssl but the server buffer is created in managed code. The problem then is openssl uses this at some future time. That means there is a race around if someone disposes SSLStream and if openssl uses that buffer which might not be pinned anymore. This was exaggerated by other problems but showed up as memory corruption.

Now there is only a slim chance but still a chance of it happening.

With this we control the selection (we are basically doing exactly the same thing as the openssl method).  But we can ensure we always return a pointer into the openssl buffer and that no matter the openssl implementation we can Provide the same behaviour.